packages/runtime/container-runtime/src/channelCollection.ts

```diff
@@ -819,84 +819,121 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
-		message: ISequencedDocumentMessage,
+	/**
+	 * Process channel type messages. The messages here are contiguous channel type messages in a batch. Bunch
+	 * of contiguous messages for a data store should be sent to it together.
+	 * @param messagesWithMetadata - A bunch of contiguous channel type messages and their metadata.
+	 * @param local - true if the messages were originally generated by the client receiving it.
+	 */
+	public processChannelMessages(
+		messagesWithMetadata: {
+			message: ISequencedDocumentMessage;
+			localOpMetadata: unknown;
+		}[],
 		local: boolean,
-		localMessageMetadata: unknown,
-	) {
-		const context = this.contexts.get(address);
-
-		// If the data store has been deleted, log an error and ignore this message. This helps prevent document
-		// corruption in case a deleted data store accidentally submitted an op.
-		if (this.checkAndLogIfDeleted(address, context, "Changed", "processFluidDataStoreOp")) {
+	): void {
+		if (messagesWithMetadata.length === 0) {
 			return;
 		}
 
-		if (context === undefined) {
-			// Former assert 0x162
-			throw DataProcessingError.create(
-				"No context for op",
-				"processFluidDataStoreOp",
-				message,
-				{
-					local,
-					messageDetails: JSON.stringify({
-						type: message.type,
-						contentType: typeof message.contents,
-					}),
-					...tagCodeArtifacts({ address }),
-				},
-			);
-		}
+		/**
+		 * Bunch contiguous messages for the same data store and send them together.
+		 * This is an optimization mainly for DDSes, where it can process a bunch of ops together. DDSes
+		 * like merge tree or shared tree can process ops more efficiently when they are bunched together.
+		 */
+		let previousAddress: string | undefined;
+		let contextMessages: {
+			message: ISequencedDocumentMessage;
+			localOpMetadata: unknown;
+		}[] = [];
+
+		messagesWithMetadata.forEach((messageWithMetadata) => {
+			const { message, localOpMetadata } = messageWithMetadata;
+			const envelope = message.contents as IEnvelope;
+			const address = envelope.address;
+			const context = this.contexts.get(address);
+
+			// If the data store has been deleted, log an error and ignore this message. This helps prevent document
+			// corruption in case a deleted data store accidentally submitted an op.
+			if (this.checkAndLogIfDeleted(address, context, "Changed", "processFluidDataStoreOp")) {
+				return;
+			}
+
+			if (context === undefined) {
+				// Former assert 0x162
+				throw DataProcessingError.create(
+					"No context for op",
+					"processFluidDataStoreOp",
+					message,
+					{
+						local,
+						messageDetails: JSON.stringify({
+							type: message.type,
+							contentType: typeof message.contents,
+						}),
+						...tagCodeArtifacts({ address }),
+					},
+				);
+			}
 
-		context.process(message, local, localMessageMetadata);
+			const innerContents = envelope.contents as FluidDataStoreMessage;
+			const contextMessage: ISequencedDocumentMessage = {
+				...message,
```

Comment:
this is what i want to avoid doing over and over again, just pass the parts, as the receiver of this object almost always splits it apart first thing.

Reply:
there are a lot of places i think we can remove the wrapping. some might make it easy to add better typing.

Reply:
Updated this as much as I could. There are still couple of places we have to do this for updating the `type`.